### PR TITLE
Avoidance Interface CI

### PR DESCRIPTION
**Pull request by mrivi**
This PR is an attempt to get an automated test on the avoidance interface going to check if any future PR would brake the interface.

Test: the vehicle will fly a mission where flying a direct line between the two waypoints leads to a collision with a wall. The avoidance maneuver is simple enough to assure that the local_planner can always find a path around the wall. If the vehicle cannot finish the mission before a timeout, the test fails. 

Current Work in Progress: I need to setup the catkin environment for the avoidance to run. All the steps are done in the script `rostest_avoidance_run.sh`. I see that in the other ROS tests the equivalent script is called [here](https://github.com/PX4/Firmware/blob/d7580aa676b502b755281a0c7a7e5780b1df9066/.ci/Jenkinsfile-SITL_tests#L218). How can I change it to run `rostest_avoidance_run.sh` for the avoidance test?

Future Work: we can use the Gazebo collision topic to detect if the vehicle has crashed

### CODE REVIEW

**Comment by lamping7** [comment]
@mrivi I did option 2. Jenkins is fine with the test now, but there seems to be other issues. Take a look at http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-SITL_tests/detail/pr-avoidance_ci_rebased/13/pipeline/82 and https://logs.px4.io/plot_app?log=69c95f98-26a4-4d87-a739-73c6baa97fc2

**Comment by mrivi** [comment]
@lamping7 thank you!!! 

I think the problem is that the avoidance is not enabled (param MPC_OBS_AVOID is not set to 1). I'll quickly try to set that in the iris config. I think we need another iris vehicle config (like those `*_iris_vision`, `*_irsi_opt_flow`) to set the parameter.

**Comment by lamping7** [comment]
I'm seeing `Laser Plugin (ns = )  <tf_prefix_>, set to ""

Aborted (core dumped)`

**Comment by mrivi** [comment]
@lamping7 @dagar avoidance test is now working! thank you for the help!
I'll put the `param set` for `MPC_OBS_AVOID` on a different target than the normal iris probably tomorrow.

**Comment by lamping7** [comment]
Awesome. The Jenkins failure is another data link lost issue.

**Comment by mrivi** [comment]
@dagar good to merge?

File: .ci/Jenkinsfile-SITL_tests [hunk 1]
```diff
@@ -127,6 +127,13 @@ pipeline {
               mission: "VTOL_mission_1",
               vehicle: "tiltrotor"
             ],
+            [
+              name: "avoidance",
```
Comment (dagar) [inline .ci/Jenkinsfile-SITL_tests]: avoidance -> MC_avoidance?

File: test/rostest_avoidance_run.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#! /bin/bash
```
Comment (dagar) [inline test/rostest_avoidance_run.sh]: This is fine for now, but we should try to think of a better way to do this. For example having it locally and running repeatedly shouldn't have to clone and build each time.

**Comment by mrivi** [comment]
@lamping7 fair points. 
- I have modified the `avoidance.plan` to include lading. This is what it does locally on my machine. Lets check if it matches the output on the ci
qgc_avoid_ci

- the obstacle is defined in this [world file](https://github.com/PX4/Firmware/blob/687ae9c6c792419da677547a2e535ce31edc7870/test/mavros_posix_test_avoidance.test#L10) that is in the avoidance repository

**Comment by lamping7** [comment]
1. Can you use 5 meters for take off alt to match other MC tests?
2. The takeoff altitude isn't the problem.
    - `[ERROR] [1541060312.830477210, 44.581000000]: FCU: Failsafe enabled: no global position`
    - `gzserver: /usr/include/boost/smart_ptr/shared_ptr.hpp:648: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::rendering::DepthCamera; typename boost::detail::sp_member_access<T>::type = gazebo::rendering::DepthCamera*]: 'Assertion px != 0' failed.`
    - I'd probably set the gazebo verbose flags in the launch file for testing to see what is going on.

**Comment by mrivi** [comment]
@mrivi before there was a mission check failure on the takeoff altitude. I can increase it but I also need to modify the obstacle. 
Let me first get the output from avoidance on the console.

**Comment by lamping7** [comment]
> @lamping7 before there was a mission check failure on the takeoff altitude. I can increase it but I also need to modify the obstacle.

Yes, I saw that, but I'm thinking it was a result of not having global position. You don't need to increase it. I was just suggesting it for consistency. I haven't pulled up the world to visualize it, so I didn't know if you could or not easily.

**Comment by TSC21** [comment]
@mrivi going to try on my side and let you know

**Comment by mrivi** [comment]
@TSC21 @lamping7 I run the contains on my laptop and retrieve the spwan log

```  [rospy.client][INFO] 2018-11-23 11:50:54,697: init_node, name[/spawn_model], pid[11572]
[xmlrpc][INFO] 2018-11-23 11:50:54,698: XML-RPC server binding to 0.0.0.0:0
[xmlrpc][INFO] 2018-11-23 11:50:54,698: Started XML-RPC server [http://9a1b26ca3e60:44181/]
[rospy.impl.masterslave][INFO] 2018-11-23 11:50:54,698: _ready: http://9a1b26ca3e60:44181/
[xmlrpc][INFO] 2018-11-23 11:50:54,698: xml rpc node: starting XML-RPC server
[rospy.registration][INFO] 2018-11-23 11:50:54,699: Registering with master node http://localhost:11311
[rospy.init][INFO] 2018-11-23 11:50:54,702: ROS Slave URI: [http://9a1b26ca3e60:44181/]
[rospy.init][INFO] 2018-11-23 11:50:54,703: registered with master
[rospy.rosout][INFO] 2018-11-23 11:50:54,703: initializing /rosout core topic
[rospy.rosout][INFO] 2018-11-23 11:50:54,705: connected to core topic /rosout
[rospy.simtime][INFO] 2018-11-23 11:50:54,712: initializing /clock core topic
[rospy.simtime][INFO] 2018-11-23 11:50:54,716: connected to core topic /clock
[rosout][INFO] 2018-11-23 11:50:54,721: Loading model XML from Gazebo Model Database
[rosout][INFO] 2018-11-23 11:50:54,721: Waiting for service /gazebo/spawn_sdf_model
[rosout][INFO] 2018-11-23 11:50:54,724: Calling service /gazebo/spawn_sdf_model
[rospy.internal][INFO] 2018-11-23 11:50:54,734: topic[/clock] adding connection to [http://9a1b26ca3e60:35775/], count 0
[rospy.internal][INFO] 2018-11-23 11:50:54,882: topic[/rosout] adding connection to [/rosout], count 0
[rosout][INFO] 2018-11-23 11:50:55,604: Spawn status: SpawnModel: Successfully spawned entity
[rospy.core][INFO] 2018-11-23 11:50:55,605: signal_shutdown [atexit]
[rospy.internal][INFO] 2018-11-23 11:50:55,609: topic[/rosout] removing connection to /rosout
[rospy.internal][INFO] 2018-11-23 11:50:55,609: topic[/clock] removing connection to http://9a1b26ca3e60:35775/
[rospy.impl.masterslave][INFO] 2018-11-23 11:50:55,609: atexit  ```

It seems that the model was successfully spawned

**Comment by TSC21** [comment]
@mrivi you are talking about the below error?
```sh
gzserver: /usr/include/boost/smart_ptr/shared_ptr.hpp:648: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::rendering::DepthCamera; typename boost::detail::sp_member_access<T>::type = gazebo::rendering::DepthCamera*]: Assertion `px != 0' failed.
```

**Comment by TSC21** [comment]
About
```sh
gzserver: /usr/include/boost/smart_ptr/shared_ptr.hpp:648: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = gazebo::rendering::DepthCamera; typename boost::detail::sp_member_access<T>::type = gazebo::rendering::DepthCamera*]: Assertion `px != 0' failed.
```

Issue a `make clean`. Then, try putting `source /usr/share/gazebo/setup.sh` in https://github.com/PX4/Firmware/blob/master/Tools/setup_gazebo.bash#L17 and then run the pipeline again to see if it works.

**Comment by mrivi** [comment]
@TSC21 both errors. `Point cloud timeout ...` is a consequence of the other error.

**Comment by TSC21** [comment]
@mrivi have you tried to run Gazebo in gdb and check the backtrace? Try passing the `debugger` argument on https://github.com/PX4/Firmware/blob/master/Tools/sitl_run.sh.

**Comment by mrivi** [comment]
@dagar I am testing container px4-dev-ros from today and I have the following errors:
```bash 
px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/test/rostest_avoidance_run.sh mavros_posix_test_avoidance.test mission:=avoidance vehicle:=iris_obs_avoid

px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/test/rostest_avoidance_run.sh: line 6: Xvfb: command not found
```

```bash
Errors << local_planner:make /tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/logs/local_planner/build.make.000.log

In file included from /tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/src/avoidance/local_planner/src/nodes/local_planner_node.cpp:1:0:

/tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/src/avoidance/local_planner/src/nodes/local_planner_node.h:12:48: fatal error: mavros_msgs/CompanionProcessStatus.h: No such file or directory

compilation terminated.

make[2]: *** [CMakeFiles/local_planner_node.dir/src/nodes/local_planner_node.cpp.o] Error 1

make[2]: *** Waiting for unfinished jobs....

In file included from /tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/src/avoidance/local_planner/src/nodes/local_planner_node_main.cpp:2:0:

/tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/src/avoidance/local_planner/src/nodes/local_planner_node.h:12:48: fatal error: mavros_msgs/CompanionProcessStatus.h: No such file or directory

compilation terminated.

make[2]: *** [CMakeFiles/local_planner_node.dir/src/nodes/local_planner_node_main.cpp.o] Error 1

make[1]: *** [CMakeFiles/local_planner_node.dir/all] Error 2

make: *** [all] Error 2

cd /tmp/jenkins/workspace/TL_tests_pr-avoidance_ci_rebased/px4-px4_sitl_default-v1.9.0-beta1-35-gfd6ed230e/px4/catkin_ws/build/local_planner; catkin build --get-env local_planner | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
```

I have built the container locally this morning and I could run `xvfb` and the avoidance (correct mavros version). Even pulling the image from Dockerhub on my machine, I see the correct mavros version (0.28) and xvfb runs without issues.

What am I doing wrong??

**Comment by lamping7** [comment]
@mrivi I think it's because your build container is different than the "run" or "testing" container (still at 2018-09-11).

**Comment by mrivi** [comment]
@lamping7 true, that was stupid :(
Now, xvfb is failing due to a permission error

```bash
+ px4-px4_sitl_default-v1.9.0-beta1-36-ga74bb362c/px4/test/rostest_avoidance_run.sh mavros_posix_test_avoidance.test mission:=avoidance vehicle:=iris_obs_avoid
_XSERVTransmkdir: ERROR: euid != 0,directory /tmp/.X11-unix will not be created
```

**Comment by lamping7** [comment]
Can we launch xvbf at the docker level?

**Comment by lamping7** [comment]
I dug into this **a lot**. I used your PR (mostly unmodified, only had to keep changing the source for ROS) to test against the three setups below:

Dockerfile_ros-kinetic (changing to gazebo 7):
- Kinetic, Gazebo 7: https://logs.px4.io/plot_app?log=3c696069-0bd8-4bb2-a503-a76f00bce29a

Dockerfile_ros-kinetic (as is):
- Kinetic, Gazebo 8: https://logs.px4.io/plot_app?log=38da9da0-a682-478c-bfda-fe6f53cfba52

Dockerfile_ros (as is):
- Melodic, Gazebo 9: https://logs.px4.io/plot_app?log=3d6e4779-c4ef-4a4f-86b9-e101642da9cc

Conclusion: They all act the exact same (failing the test the same way -- essentially not avoiding very well at all) and all worked with Xvfb the way I set it up, slightly different than your PR.

Details:
I tested by building the images locally, and spinning up the containers without any special args (only `-it`). Inside the container I ran:
1. git clone https://github.com/PX4/Firmware.git
2. cd Firmware && git checkout pr-avoidance_ci_rebased
3. make px4_sitl_default && make px4_sitl_default sitl_gazebo
4. test/rostest_avoidance_run.sh mavros_posix_test_avoidance.test mission:=avoidance

I'll link to the Xvfb setup when it get into the containers repo. Do you have this working **inside** of the PX4 containers locally?

**Comment by dagar** [comment]
We need avoidance to be versioned, which likely means building it in a container that's tagged.

This could be done with a Dockerfile that lives in https://github.com/PX4/avoidance.

**Comment by dagar** [comment]
Can we get this current branch squashed and rebased on master?

**Comment by mrivi** [comment]
@dagar we will have a stable version of avoidance sometimes this week

**Comment by lamping7** [comment]
https://github.com/PX4/Firmware/issues/11523

**Comment by lamping7** [comment]
Getting there... rebase after https://github.com/PX4/Firmware/pull/11606 if everything checks out. Seems promising on the SITL side.

**Comment by mrivi** [comment]
The simulation is now working. The remaining problem is that the avoidance doesn't let the vehicle land. There are two possible solutions:
- disable the avoidance setpoint when the current triplet is type SETPOINT_TYPE_LAND or when the nav state is AUTO_LOITER on the Firmware side
- fix the issue on the avoidance side

**Comment by mrivi** [comment]
@dagar the avoidance test is now working against the stable avoidance release.
Anything else needed before merging?

**Comment by mrivi** [comment]
@dagar can we merge this?

File: integrationtests/python_src/px4_it/mavros/missions/avoidance.plan [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        "version": 2
+    },
+    "groundStation": "QGroundControl",
+    "mission": {
+        "cruiseSpeed": 15,
+        "firmwareType": 12,
+        "hoverSpeed": 5,
+        "items": [
+            {
+                "AMSLAltAboveTerrain": 4,
```
Comment (lamping7) [inline integrationtests/python_src/px4_it/mavros/missions/avoidance.plan]: I've never had to set this. To help me out, why is this needed? Not suggesting a change. Just curious.

File: test/rostest_avoidance_run.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+PX4_SRC_DIR=${DIR}/..
+
+source /opt/ros/${ROS_DISTRO}/setup.bash
```
Comment (lamping7) [inline test/rostest_avoidance_run.sh]: This is good for Kinetic (Gazebo 7) and Melodic (Gazebo 9)? Using Kinetic for now, so maybe change to:
`source /opt/ros/${ROS_DISTRO:-kinetic}/setup.bash`, matching https://github.com/PX4/Firmware/blob/2ae2331c15e10008f01c55d90f549cae91716acb/test/rostest_px4_run.sh#L6

**Comment by lamping7** [comment]
Also, do we want to cut down the messages spamming the console multiple times a second?
```
[ INFO] [1552481031.530698617, 33.248000000]: [SP]Tree (72 nodes, 2 path nodes, 10 expanded) calculated in 6.55ms.
[ INFO] [1552481031.553791537, 33.272000000]: [WG] Speed adapted WP: [10.778049 1.757204 4.033766].
[ INFO] [1552481031.589228580, 33.304000000]: [WG] Speed adapted WP: [10.777305 1.752048 4.026687].
[ INFO] [1552481031.621368949, 33.340000000]: [WG] Speed adapted WP: [10.779245 1.758199 4.021173].
[ INFO] [1552481031.653143228, 33.372000000]: [WG] Speed adapted WP: [10.167651 2.287094 4.085506].
[ INFO] [1552481031.653242088, 33.372000000]: [OA] Planning started, using 3 cameras
```

File: integrationtests/python_src/px4_it/mavros/missions/avoidance.plan (hunk 3 above)
Comment (mrivi) [reply integrationtests/python_src/px4_it/mavros/missions/avoidance.plan]: Not sure why it was set that way. I downloaded the mission plan I was using to run the test locally. I reverted back to nan.

**Comment by mrivi** [comment]
@lamping7 I addressed your comments. 

I killed all the avoidance outputs from the console.
Yes, the warning on the third waypoint is due to the relative height being set to -1. This is necessary to make the vehicle land and disarm. Once the avoidance can handle landing correctly we can change the relative altiude back to 0.

**Comment by davids5** [comment]
dev call discussion:Get a clear failure message FYI: @mrivi

**Comment by mrivi** [comment]
I rebased because I realized that the `OBS_AVOID` parameter wasn't up to date. It switched from `MPC_OBS_AVOID` to `COM_OBS_AVOID` in #11550

**Comment by nicovanduijn** [comment]
If we want to merge our efforts, you could cherry-pick 70c9c78dc9fb27a509d11ca718890ff72b187153

**Comment by mrivi** [comment]
It has been long six months to get this PR working so I am not going to change anything at the moment. This is a good first step and I would like to keep the improvements in a separate PR.
@dagar could you please review it? I really would like to see this merged rather sooner than later.

**Comment by nicovanduijn** [comment]
Ok, I'll work on top of this then. Let's get it in

**Comment by dagar** [comment]
Thanks everyone!
